<a href="https://colab.research.google.com/github/anabiarochar/Mestrado_AlgoritmosProgramacao/blob/main/Projeto1/2_Sele%C3%A7%C3%A3o.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Projeto 1 - Algoritmo de Ordenação Seleção (Selection Sort)

Abaixo estão os algorítmos para ordenação, utilizando o método seleção de ordenação, com entradas de vetores de tamanhos 1.000, 10.000 e 100.000
- aleatórias
- ordenadas em forma crescente
- ordenadas em forma decrescente

Abaixo de cada algorítmo estão os tempos observados somando 4 tentativas.

##Vetor em ordem aleatória com 1000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 0.033886 segundos
[1, 3, 8, 9, 9, 10, 10, 10, 10, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.025019 segundos  
Tentativa 2 - Tempo de execução: 0.047944 segundos  
Tentativa 3 - Tempo de execução: 0.051149 segundos  
Tentativa 4 - Tempo de execução: 0.025113 segundos  

##Vetor em ordem aleatória com 10000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 2.587237 segundos
[3, 3, 4, 8, 9, 9, 9, 11, 12, 12]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 2.562468 segundos  
Tentativa 2 - Tempo de execução: 2.686351 segundos  
Tentativa 3 - Tempo de execução: 2.500412 segundos  
Tentativa 4 - Tempo de execução: 3.145576 segundos  

##Vetor em ordem aleatória com 100000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_aleatorio.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 383.648474 segundos
[2, 5, 6, 6, 8, 9, 10, 13, 14, 14]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 316.358742 segundos  
Tentativa 2 - Tempo de execução: 383.648474 segundos  
Tentativa 3 - Tempo de execução:  
Tentativa 4 - Tempo de execução:  

##Vetor em ordem crescente com 1000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 0.027577 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.059829 segundos   
Tentativa 2 - Tempo de execução: 0.029312 segundos  
Tentativa 3 - Tempo de execução: 0.028649 segundos  
Tentativa 4 - Tempo de execução: 0.044217 segundos

##Vetor em ordem crescente com 10000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 2.569358 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 3.956059 segundos   
Tentativa 2 - Tempo de execução: 2.595252 segundos  
Tentativa 3 - Tempo de execução: 2.628583 segundos  
Tentativa 4 - Tempo de execução: 2.557347 segundos

##Vetor em ordem crescente com 100000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_crescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 291.600287 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 286.392151 segundos     
Tentativa 2 - Tempo de execução: 290.976181 segundos  
Tentativa 3 - Tempo de execução: 282.764232 segundos  
Tentativa 4 - Tempo de execução: 291.600287 segundos

##Vetor em ordem decrescente com 1000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_1000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 0.025376 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 0.027707 segundos   
Tentativa 2 - Tempo de execução: 0.050121 segundos  
Tentativa 3 - Tempo de execução: 0.024865 segundos  
Tentativa 4 - Tempo de execução: 0.025376 segundos

##Vetor em ordem decrescente com 10000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_10000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 2.615781 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 4.169898 segundos   
Tentativa 2 - Tempo de execução: 3.692997 segundos  
Tentativa 3 - Tempo de execução: 2.691385 segundos  
Tentativa 4 - Tempo de execução: 2.644515 segundos

##Vetor em ordem decrescente com 100000 elementos:


In [ ]:
import requests
import time

# Importar dados
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/main/Projeto1/Vetores/vetor_100000_decrescente.txt"

try:
    response = requests.get(url)
    content = response.text

    # Ler arquivo como vetor separado por vírgula
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")


# Método de Seleção
def selectionSort(vetor):
  n = len(vetor)
  for i in range(n):
      min = i
      for j in range(i + 1, n):
            if vetor[j] < vetor[min]:
                min = j

      vetor[i], vetor[min] = vetor[min], vetor[i]

  return vetor

# Execução e Tempo
if vetor:
    inicio = time.perf_counter()
    vetor_ordenado = selectionSort(vetor)
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(vetor[:10])

else:
    print("Vetor vazio")



Tempo de execução: 302.118604 segundos
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


###Tempo de Execução:

Tentativa 1 - Tempo de execução: 292.599626 segundos  
Tentativa 2 - Tempo de execução: 298.507569 segundos  
Tentativa 3 - Tempo de execução: 293.371434 segundos  
Tentativa 4 - Tempo de execução: 302.118604 segundos  

##Resumo dos Tempos de Execução

| Caso | n | Tentativa 1 (s) | Tentativa 2 (s) | Tentativa 3 (s) | Tentativa 4 (s) | Média (s) |
|---|---:|---:|---:|---:|---:|---:|
| Aleatório | 1000 | 0.025019 | 0.047944 | 0.051149 | 0.025113 | 0.037306 |
| Aleatório | 10000 | 2.562468 | 2.686351 | 2.500412 | 3.145576 | 2.723702 |
| Aleatório | 100000 | 316.358742 | 383.648474 | — | — | 350.003608*\ |
| Crescente | 1000 | 0.059829 | 0.029312 | 0.028649 | 0.044217 | 0.040502 |
| Crescente | 10000 | 3.956059 | 2.595252 | 2.628583 | 2.557347 | 2.934310 |
| Crescente | 100000 | 286.392151 | 290.976181 | 282.764232 | 291.600287 | 287.933213 |
| Decrescente | 1000 | 0.027707 | 0.050121 | 0.024865 | 0.025376 | 0.032017 |
| Decrescente | 10000 | 4.169898 | 3.692997 | 2.691385 | 2.644515 | 3.299699 |
| Decrescente | 100000 | 292.599626 | 298.507569 | 293.371434 | 302.118604 | 296.649308 |

\* No caso "Aleatório / 100000", a média foi calculada com 2 tentativas preenchidas (as tentativas 3 e 4 estão em branco no notebook).